File: class22/python/PES2203-ComunicacaoServicos.py
"""
PES2203 - Comunicação Síncrona x Assíncrona entre Serviços
Aula 22: Microsserviços e Arquiteturas Distribuídas
Projeto e Engenharia de Software - Prof. Giovane Barcelos

Contrasta os dois estilos de comunicação entre microsserviços:
- Síncrona (request/response): quem chama espera a resposta antes de continuar.
- Assíncrona (mensageria/eventos): quem chama publica um evento e segue
  em frente, sem esperar quem vai consumi-lo nem quando.
"""

import queue


# --- COMUNICAÇÃO SÍNCRONA: Pedidos chama Estoque e espera a resposta ---
class ServicoEstoqueSincrono:
    def consultar_estoque_sincrono(self, placa: str) -> int:
        print(f"[Estoque] Consultando quantidade de {placa}...")
        return 3  # resposta imediata, quem chamou fica bloqueado até aqui


class ServicoPedidosSincrono:
    def __init__(self, estoque: ServicoEstoqueSincrono):
        self.estoque = estoque

    def criar_pedido(self, placa: str) -> str:
        quantidade = self.estoque.consultar_estoque_sincrono(placa)  # bloqueia
        print(f"[Pedidos] Resposta recebida: {quantidade} unidades")
        return "Criado" if quantidade > 0 else "Recusado"


# --- COMUNICAÇÃO ASSÍNCRONA: Pedidos publica um evento e não espera ---
class BarramentoEventos:
    """Simula um message broker (ex.: RabbitMQ/Kafka) com uma fila em memória."""

    def __init__(self):
        self._fila = queue.Queue()
        self._assinantes = []

    def assinar(self, callback) -> None:
        self._assinantes.append(callback)

    def publicar(self, evento: str, dados: dict) -> None:
        print(f"[Barramento] Evento publicado: {evento} {dados}")
        self._fila.put((evento, dados))

    def processar_fila(self) -> None:
        while not self._fila.empty():
            evento, dados = self._fila.get()
            for callback in self._assinantes:
                callback(evento, dados)


class ServicoPedidosAssincrono:
    def __init__(self, barramento: BarramentoEventos):
        self.barramento = barramento

    def notificar_pedido_criado(self, placa: str) -> None:
        print(f"[Pedidos] Publicando evento e seguindo em frente (não bloqueia)...")
        self.barramento.publicar("PedidoCriado", {"placa": placa})


class ServicoEmailAssincrono:
    def ao_receber_evento(self, evento: str, dados: dict) -> None:
        if evento == "PedidoCriado":
            print(f"[E-mail] Recebeu o evento com atraso e envia confirmação para {dados['placa']}")


if __name__ == "__main__":
    print("=" * 60)
    print("  COMUNICAÇÃO SÍNCRONA x ASSÍNCRONA")
    print("=" * 60)

    print("\n--- Síncrona (request/response) ---")
    pedidos_sinc = ServicoPedidosSincrono(ServicoEstoqueSincrono())
    resultado = pedidos_sinc.criar_pedido("ABC1234")
    print(f"  Resultado: {resultado}")
    assert resultado == "Criado"

    print("\n--- Assíncrona (mensageria/eventos) ---")
    barramento = BarramentoEventos()
    email = ServicoEmailAssincrono()
    barramento.assinar(email.ao_receber_evento)

    pedidos_assinc = ServicoPedidosAssincrono(barramento)
    pedidos_assinc.notificar_pedido_criado("XYZ9876")
    print("  [Pedidos] Já retornou para o cliente, sem esperar o e-mail ser enviado.")

    barramento.processar_fila()  # o consumo do evento acontece depois, de forma desacoplada

    print("\n✓ Síncrona: quem chama fica bloqueado esperando a resposta.")
    print("✓ Assíncrona: quem chama publica e segue, o consumidor processa depois.")

File: class22/java/PES2203-ComunicacaoServicos.java
/*
 * PES2203 - Comunicação Síncrona x Assíncrona entre Serviços
 * Aula 22: Microsserviços e Arquiteturas Distribuídas
 * Projeto e Engenharia de Software - Prof. Giovane Barcelos
 *
 * Contrasta os dois estilos de comunicação entre microsserviços:
 * - Síncrona (request/response): quem chama espera a resposta antes de continuar.
 * - Assíncrona (mensageria/eventos): quem chama publica um evento e segue
 *   em frente, sem esperar quem vai consumi-lo nem quando.
 *
 * Compilar/executar:
 *   javac PES2203-ComunicacaoServicos.java && java ComunicacaoServicos
 */

import java.util.*;

// --- COMUNICAÇÃO SÍNCRONA: Pedidos chama Estoque e espera a resposta ---
class ServicoEstoqueSincrono {
    int consultarEstoqueSincrono(String placa) {
        System.out.println("[Estoque] Consultando quantidade de " + placa + "...");
        return 3; // resposta imediata, quem chamou fica bloqueado até aqui
    }
}

class ServicoPedidosSincrono {
    private final ServicoEstoqueSincrono estoque;

    ServicoPedidosSincrono(ServicoEstoqueSincrono estoque) {
        this.estoque = estoque;
    }

    String criarPedido(String placa) {
        int quantidade = estoque.consultarEstoqueSincrono(placa); // bloqueia
        System.out.println("[Pedidos] Resposta recebida: " + quantidade + " unidades");
        return quantidade > 0 ? "Criado" : "Recusado";
    }
}

// --- COMUNICAÇÃO ASSÍNCRONA: Pedidos publica um evento e não espera ---
interface OuvinteEvento {
    void aoReceberEvento(String evento, Map<String, String> dados);
}

// Simula um message broker (ex.: RabbitMQ/Kafka) com uma fila em memória.
class BarramentoEventos {
    private final Queue<Map.Entry<String, Map<String, String>>> fila = new LinkedList<>();
    private final List<OuvinteEvento> assinantes = new ArrayList<>();

    void assinar(OuvinteEvento ouvinte) {
        assinantes.add(ouvinte);
    }

    void publicar(String evento, Map<String, String> dados) {
        System.out.println("[Barramento] Evento publicado: " + evento + " " + dados);
        fila.add(new AbstractMap.SimpleEntry<>(evento, dados));
    }

    void processarFila() {
        while (!fila.isEmpty()) {
            Map.Entry<String, Map<String, String>> item = fila.poll();
            for (OuvinteEvento ouvinte : assinantes) {
                ouvinte.aoReceberEvento(item.getKey(), item.getValue());
            }
        }
    }
}

class ServicoPedidosAssincrono {
    private final BarramentoEventos barramento;

    ServicoPedidosAssincrono(BarramentoEventos barramento) {
        this.barramento = barramento;
    }

    void notificarPedidoCriado(String placa) {
        System.out.println("[Pedidos] Publicando evento e seguindo em frente (não bloqueia)...");
        Map<String, String> dados = new HashMap<>();
        dados.put("placa", placa);
        barramento.publicar("PedidoCriado", dados);
    }
}

class ServicoEmailAssincrono implements OuvinteEvento {
    @Override
    public void aoReceberEvento(String evento, Map<String, String> dados) {
        if (evento.equals("PedidoCriado")) {
            System.out.println("[E-mail] Recebeu o evento com atraso e envia confirmação para " + dados.get("placa"));
        }
    }
}

class ComunicacaoServicos {
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("  COMUNICAÇÃO SÍNCRONA x ASSÍNCRONA");
        System.out.println("=".repeat(60));

        System.out.println("\n--- Síncrona (request/response) ---");
        ServicoPedidosSincrono pedidosSinc = new ServicoPedidosSincrono(new ServicoEstoqueSincrono());
        String resultado = pedidosSinc.criarPedido("ABC1234");
        System.out.println("  Resultado: " + resultado);
        assert resultado.equals("Criado");

        System.out.println("\n--- Assíncrona (mensageria/eventos) ---");
        BarramentoEventos barramento = new BarramentoEventos();
        barramento.assinar(new ServicoEmailAssincrono());

        ServicoPedidosAssincrono pedidosAssinc = new ServicoPedidosAssincrono(barramento);
        pedidosAssinc.notificarPedidoCriado("XYZ9876");
        System.out.println("  [Pedidos] Já retornou para o cliente, sem esperar o e-mail ser enviado.");

        barramento.processarFila(); // o consumo do evento acontece depois, de forma desacoplada

        System.out.println("\n✓ Síncrona: quem chama fica bloqueado esperando a resposta.");
        System.out.println("✓ Assíncrona: quem chama publica e segue, o consumidor processa depois.");
    }
}
